File: py/aoc2022/day19.py
"""
Day 19: Not Enough Minerals
"""

import itertools
import math
import re
import sys
from collections import defaultdict
from concurrent.futures import ProcessPoolExecutor

SAMPLE_INPUT = [
    "Blueprint 1: Each ore robot costs 4 ore. "
    "Each clay robot costs 2 ore. "
    "Each obsidian robot costs 3 ore and 14 clay. "
    "Each geode robot costs 2 ore and 7 obsidian.",
    "Blueprint 2: Each ore robot costs 2 ore. "
    "Each clay robot costs 3 ore. "
    "Each obsidian robot costs 3 ore and 8 clay. "
    "Each geode robot costs 3 ore and 12 obsidian.",
]

ID_PATTERN = re.compile(r"Blueprint (\d+):")
PART_PATTERN = re.compile(r"Each (\w+) robot ([^.]+)[.]")
COST_PATTERN = re.compile(r"(?:costs | and )(\d+) (\w+)")


def _parse(lines):
    for line in lines:
        yield int(re.match(ID_PATTERN, line).group(1)), {
            part_match.group(1): defaultdict(
                int,
                {
                    cost_match.group(2): int(cost_match.group(1))
                    for cost_match in re.finditer(COST_PATTERN, part_match.group(2))
                },
            )
            for part_match in re.finditer(PART_PATTERN, line)
        }


def _potential(blueprint, robots, materials, time):
    additional_robots = defaultdict(int)
    potential_materials = defaultdict(int, materials)
    for _ in range(time):
        for robot in blueprint:
            potential_materials[robot] += robots[robot] + additional_robots[robot]
        for robot, costs in blueprint.items():
            if all(
                potential_materials[material] >= cost * (additional_robots[robot] + 1)
                for material, cost in costs.items()
            ):
                additional_robots[robot] += 1
    return potential_materials["geode"]


def _geodes(blueprint, time):
    # pylint: disable=too-many-locals
    max_values = {}
    for costs in blueprint.values():
        for material, cost in costs.items():
            max_values[material] = max(max_values.get(material, 0), cost)
    best, queue = 0, [(defaultdict(int, {"ore": 1}), defaultdict(int), time)]
    while queue:
        robots, materials, time = queue.pop()
        if _potential(blueprint, robots, materials, time) < best:
            continue
        estimate = materials["geode"] + robots["geode"] * time
        if estimate > best:
            best = estimate
        for robot, costs in blueprint.items():
            if robot in max_values and robots[robot] >= max_values[robot]:
                continue
            delta = max(
                0
                if (demand := costs[material] - materials[material]) <= 0
                else math.inf
                if (supply := robots[material]) <= 0
                else (demand + supply - 1) // supply
                for material in blueprint
            )
            if delta < time:
                new_robots = defaultdict(int, robots)
                new_robots[robot] += 1
                new_materials = defaultdict(int, materials)
                for material, count in robots.items():
                    new_materials[material] += count * (delta + 1)
                for material, cost in costs.items():
                    new_materials[material] -= cost
                queue.append((new_robots, new_materials, time - delta - 1))
    return best


def _part1_worker(blueprint):
    bid, blueprint = blueprint
    geodes = _geodes(blueprint, 24)
    print((bid, geodes), file=sys.stderr)
    return bid * geodes


def part1(lines):
    """
    >>> part1(SAMPLE_INPUT)
    33
    """
    with ProcessPoolExecutor() as executor:
        return sum(executor.map(_part1_worker, _parse(lines)))


def _part2_worker(blueprint):
    bid, blueprint = blueprint
    geodes = _geodes(blueprint, 32)
    print((bid, geodes), file=sys.stderr)
    return geodes


def part2(lines):
    """
    >>> part2(SAMPLE_INPUT)
    3472
    """

    with ProcessPoolExecutor() as executor:
        return math.prod(
            executor.map(_part2_worker, itertools.islice(_parse(lines), 3))
        )


parts = (part1, part2)

File: rs/src/day19.rs
use rayon::iter::{IntoParallelIterator, ParallelIterator};
use std::cmp::max;
use std::collections::BTreeMap;

type Blueprint<'a> = BTreeMap<&'a str, BTreeMap<&'a str, usize>>;

fn parse(line: &str) -> Option<(usize, Blueprint)> {
    let line = line.strip_prefix("Blueprint ")?;
    let (id, line) = line.split_once(':')?;
    let id = id.parse().ok()?;
    let blueprint = line
        .split('.')
        .filter(|part| !part.is_empty())
        .map(|part| {
            let part = part.strip_prefix(" Each ")?;
            let (robot, part) = part.split_once(' ')?;
            let part = part.strip_prefix("robot costs ")?;
            let costs = part
                .split(" and ")
                .map(|s| {
                    let (cost, material) = s.split_once(' ')?;
                    Some((material, cost.parse().ok()?))
                })
                .collect::<Option<_>>()?;
            Some((robot, costs))
        })
        .collect::<Option<_>>()?;
    Some((id, blueprint))
}

struct State<'a> {
    robots: BTreeMap<&'a str, usize>,
    materials: BTreeMap<&'a str, usize>,
    time: usize,
}

fn geodes(blueprint: &Blueprint, time: usize) -> usize {
    let mut max_values = BTreeMap::new();
    for costs in blueprint.values() {
        for (&material, &cost) in costs {
            max_values
                .entry(material)
                .and_modify(|e| *e = max(*e, cost))
                .or_insert(cost);
        }
    }
    let mut best = 0;
    let mut queue = vec![State {
        robots: [("ore", 1)].into(),
        materials: [].into(),
        time,
    }];
    while let Some(State {
        robots,
        materials,
        time,
    }) = queue.pop()
    {
        let potential = {
            let mut additional_robots = BTreeMap::new();
            let mut materials = materials.clone();
            for _ in 0..time {
                for (&robot, &count) in robots.iter().chain(additional_robots.iter()) {
                    materials
                        .entry(robot)
                        .and_modify(|e| *e += count)
                        .or_insert(count);
                }
                for robot in blueprint
                    .iter()
                    .filter(|(&robot, costs)| {
                        costs.iter().all(|(&material, &cost)| {
                            materials.get(material).copied().unwrap_or(0)
                                >= cost * (additional_robots.get(robot).unwrap_or(&0) + 1)
                        })
                    })
                    .map(|(&robot, _)| robot)
                    .collect::<Vec<_>>()
                {
                    additional_robots
                        .entry(robot)
                        .and_modify(|e| *e += 1)
                        .or_insert(1);
                }
            }
            materials.get("geode").copied().unwrap_or(0)
        };
        if potential < best {
            continue;
        }
        let estimate =
            materials.get("geode").unwrap_or(&0) + robots.get("geode").unwrap_or(&0) * time;
        best = max(best, estimate);
        for (&robot, costs) in blueprint {
            if max_values
                .get(robot)
                .filter(|&max_value| robots.get(robot).unwrap_or(&0) >= max_value)
                .is_some()
            {
                continue;
            }
            let delta = costs
                .iter()
                .filter_map(|(&material, &cost)| -> Option<usize> {
                    let demand = cost.checked_sub(*materials.get(material).unwrap_or(&0))?;
                    let supply = robots.get(material).copied().unwrap_or(0);
                    if supply == 0 {
                        Some(usize::MAX)
                    } else {
                        Some((demand + supply - 1) / supply)
                    }
                })
                .max()
                .unwrap_or(0);
            if delta < time {
                let mut materials = materials.clone();
                for (&robot, &count) in &robots {
                    materials
                        .entry(robot)
                        .and_modify(|e| *e += count * (delta + 1))
                        .or_insert(count * (delta + 1));
                }
                for (&material, &cost) in costs {
                    *materials.get_mut(material).unwrap() -= cost;
                }
                let mut robots = robots.clone();
                robots.entry(robot).and_modify(|e| *e += 1).or_insert(1);
                queue.push(State {
                    robots,
                    materials,
                    time: time - delta - 1,
                });
            }
        }
    }
    best
}

pub fn part1<'a, I, S>(lines: I) -> usize
where
    I: IntoIterator<Item = &'a S>,
    S: AsRef<str> + 'a,
{
    lines
        .into_iter()
        .filter_map(|line| parse(line.as_ref()))
        .collect::<Vec<_>>()
        .into_par_iter()
        .map(|(id, blueprint)| id * geodes(&blueprint, 24))
        .sum()
}

pub fn part2<'a, I, S>(lines: I) -> usize
where
    I: IntoIterator<Item = &'a S>,
    S: AsRef<str> + 'a,
{
    lines
        .into_iter()
        .take(3)
        .filter_map(|line| parse(line.as_ref()))
        .collect::<Vec<_>>()
        .into_par_iter()
        .map(|(_, blueprint)| geodes(&blueprint, 32))
        .product()
}

#[cfg(test)]
mod tests {
    use super::*;
    use pretty_assertions::assert_eq;

    static EXAMPLE: &[&str] = &[
        "Blueprint 1: \
            Each ore robot costs 4 ore. \
            Each clay robot costs 2 ore. \
            Each obsidian robot costs 3 ore and 14 clay. \
            Each geode robot costs 2 ore and 7 obsidian.",
        "Blueprint 2: \
            Each ore robot costs 2 ore. \
            Each clay robot costs 3 ore. \
            Each obsidian robot costs 3 ore and 8 clay. \
            Each geode robot costs 3 ore and 12 obsidian.",
    ];

    #[test]
    fn part1_examples() {
        assert_eq!(33, part1(EXAMPLE));
    }

    #[test]
    fn part2_examples() {
        assert_eq!(56 * 62, part2(EXAMPLE));
    }
}
